### PR TITLE
feat(format): add format diagnostic

### PR DESCRIPTION
**Pull request by fkuner**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

It's very difficult to find the problem when loading large files and error happens.

We hope to diagnose where the problem locates when loading formats.

For example:

```
insert into a(a,b,c) format CSV '2023-04-08 01:01:01', "Hello", 12345678,1
```

The error msg is like this:

```
Code: 1046, displayText = Parse csv error at line 0
Error occurs at row: 0:
	Column: 0, Name: a, Type: Timestamp, Parsed text: "<SINGLE QUOTE>2023-04-08 01:01:01<SINGLE QUOTE>"
	Column: 1, Name: b, Type: String, Parsed text: "<DOUBLE QUOTE>Hello<DOUBLE QUOTE>"
	Column: 2, Name: c, Type: Int32, Parsed text: "12345678"
	Error: There is no line feed. "," found instead.
.
```

So, we can find the problem quickly.

### CODE REVIEW

**Comment by PsiACE** [comment]
after https://github.com/datafuselabs/databend/pull/6443 

formats is available as a common crate

**Comment by fkuner** [comment]
> 

Got it. I will merge main in local.

**Comment by fkuner** [comment]
still need file name and row num in `streaming_load`

**Comment by sundy-li** [comment]
> still need file name and row num in `streaming_load`

we already have them and should pass the extra information into the deserializer.

File: common/formats/src/format_csv.rs [hunk 1]
```diff
@@ -215,51 +235,26 @@ impl InputFormat for CsvInputFormat {
...
-                {
-                    return Err(ErrorCode::BadBytes(format!(
-                        "Parse csv error at line {}",
-                        row_index
-                    )));
-                }
-
-                // \r\n
-                checkpoint_reader.ignore_white_spaces_and_byte(b'\n')?;
+            checkpoint_reader.push_checkpoint();
```
Comment (sundy-li) [inline common/formats/src/format_csv.rs:244]: we should pop_checkpoint if read_row works fine.

Comment (fkuner) [reply common/formats/src/format_csv.rs:244]: Yes, I forgot it.

File: common/formats/src/format_diagnostic.rs [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        &self,
+        buf: &[u8],
+        file_name: &Option<String>,
+        row_index: usize,
+        schema: DataSchemaRef,
+        min_accepted_rows: usize,
+        settings: FormatSettings,
+    ) -> Result<String> {
+        let mut out = String::new();
+        match file_name {
```
Comment (zhang2014) [inline common/formats/src/format_diagnostic.rs:73]: maybe format!() is better(without unwrap)?

File: common/io/src/buffer/buffer_read_ext.rs [hunk 3]
```diff
@@ -29,6 +29,38 @@ pub trait BufferReadExt: BufferRead {
 
     fn keep_read(&mut self, buf: &mut Vec<u8>, f: impl Fn(u8) -> bool) -> Result<usize>;
 
+    fn position(&mut self) -> Result<u8> {
+        let available = self.fill_buf()?;
+        if available.is_empty() {
+            todo!()
```
Comment (zhang2014) [inline common/io/src/buffer/buffer_read_ext.rs:35]: need to implement?

File: query/src/servers/http/v1/parallel_format_source.rs [hunk 4]
```diff
@@ -137,7 +137,12 @@ impl MultipartWorker for ParallelMultipartWorker {
                                         continue 'read;
                                     }
 
+                                    let mut read_row_num = 0;
+
                                     while !buf_slice.is_empty() {
+                                        self.input_format
+                                            .set_state(&mut state, filename.clone(), read_row_num)
+                                            .unwrap();
```
Comment (zhang2014) [inline query/src/servers/http/v1/parallel_format_source.rs:145]: need remove unwrap.

File: query/src/servers/http/v1/parallel_format_source.rs [hunk 5]
```diff
@@ -148,6 +153,9 @@ impl MultipartWorker for ParallelMultipartWorker {
                                                 }
                                             };
 
+                                        read_row_num +=
+                                            self.input_format.read_row_num(&mut state).unwrap();
```
Comment (zhang2014) [inline query/src/servers/http/v1/parallel_format_source.rs:157]: remove `unwrap`

File: common/io/src/buffer/nested_checkpoint_reader.rs [hunk 6]
```diff
@@ -60,6 +60,17 @@ impl<R: BufferRead> NestedCheckpointReader<R> {
...
     }
 
+    pub fn get_checkpoint_buffer_end(&mut self) -> &[u8] {
+        let cp = self.checkpoints[self.checkpoints.len() - 1];
+        self.fill_buf().unwrap();
+        &self.buffer[cp..self.buffer.len()]
+    }
+
+    pub fn get_top_checkpoint_pos(&self) -> usize {
+        assert!(self.checkpointing);
```
Comment (zhang2014) [inline common/io/src/buffer/nested_checkpoint_reader.rs:70]: debug_assert or check and return Result::Err.

File: common/formats/src/format.rs [hunk 7]
```diff
@@ -28,9 +31,31 @@ pub trait InputFormat: Send + Sync {
...
+        &self,
+        _checkpoint_reader: &mut NestedCheckpointReader<MemoryReader>,
+        _deserializers: &mut Vec<TypeDeserializerImpl>,
+        _row_index: usize,
+    ) -> Result<()> {
+        unimplemented!()
+    }
+
+    fn read_row_num(&self, _state: &mut Box<dyn InputState>) -> Result<usize> {
+        unimplemented!()
```
Comment (zhang2014) [inline common/formats/src/format.rs:59]: ErrorCode::unimplmented is better.

File: common/formats/src/format_diagnostic.rs (hunk 2 above)
Comment (fkuner) [reply common/formats/src/format_diagnostic.rs:73]: make sense

Comment (fkuner) [reply common/formats/src/format_diagnostic.rs:73]: `format!` has extra allocation.

File: common/formats/src/format_csv.rs [hunk 8]
```diff
@@ -311,4 +353,126 @@ impl InputFormat for CsvInputFormat {
...
+
+    fn read_row_num(&self, state: &mut Box<dyn InputState>) -> Result<usize> {
+        let state = state.as_any().downcast_mut::<CsvInputState>().unwrap();
+        Ok(state.accepted_rows)
+    }
+}
+
+#[allow(clippy::format_push_string)]
+impl FormatDiagnostic for CsvInputFormat {
+    fn deserialize_field_and_print_diagnositc_info(
```
Comment (sundy-li) [inline common/formats/src/format_csv.rs:365]: I think this function has less relationship with the format. So we need to implement it just once.

File: common/formats/src/format_tsv.rs [hunk 9]
```diff
@@ -271,4 +314,117 @@ impl InputFormat for TsvInputFormat {
...
+
+    fn read_row_num(&self, state: &mut Box<dyn InputState>) -> Result<usize> {
+        let state = state.as_any().downcast_mut::<TsvInputState>().unwrap();
+        Ok(state.accepted_rows)
+    }
+}
+
+#[allow(clippy::format_push_string)]
+impl FormatDiagnostic for TsvInputFormat {
+    fn deserialize_field_and_print_diagnositc_info(
```
Comment (sundy-li) [inline common/formats/src/format_tsv.rs:326]: ditto

File: common/formats/src/format_csv.rs (hunk 8 above)
Comment (fkuner) [reply common/formats/src/format_csv.rs:365]: make sense

**Review by zhang2014 — APPROVED** [review]

**Review by sundy-li — APPROVED** [review]

**Pull request merged**